Провести дисперсионный анализ для определения того, есть ли различия среднего роста среди взрослых футболистов, хоккеистов и штангистов.
Даны значения роста в трех группах случайно выбранных спортсменов:

Футболисты: 173, 175, 180, 178, 177, 185, 183, 182.
Хоккеисты: 177, 179, 180, 188, 177, 172, 171, 184, 180.
Штангисты: 172, 173, 169, 177, 166, 180, 178, 177, 172, 166, 170.

x1  - Футболисты: 173, 175, 180, 178, 177, 185, 183, 182.
x2  - Хоккеисты: 177, 179, 180, 188, 177, 172, 171, 184, 180.
x3  - Штангисты: 172, 173, 169, 177, 166, 180, 178, 177, 172, 166, 170.

In [27]:
import numpy as np
from scipy import stats
import matplotlib.pyplot as plt

x1 = np.array([173, 175, 180, 178, 177, 185, 183, 182])
x2 = np.array([177, 179, 180, 188, 177, 172, 171, 184, 180])
x3 = np.array([172, 173, 169, 177, 166, 180, 178, 177, 172, 166, 170])
k = 3
n = 28
# находим средние значения по выборкам
x_mean_1 = np.mean(x1)
x_mean_1

179.125

In [28]:
x_mean_2 = np.mean(x2)
x_mean_2

178.66666666666666

In [29]:
x_mean_3 = np.mean(x3)
x_mean_3 

172.72727272727272

In [30]:
total = np.concatenate([x1, x2, x3])
total

array([173, 175, 180, 178, 177, 185, 183, 182, 177, 179, 180, 188, 177,
       172, 171, 184, 180, 172, 173, 169, 177, 166, 180, 178, 177, 172,
       166, 170])

In [32]:
x_mean_total = np.mean(total)
x_mean_total 

176.46428571428572

In [33]:
# теперь расчитываем сумму квадратов отклонений от общего среднего
s = np.sum((total - x_mean_total)**2)
s

830.9642857142854

In [35]:
# сумма квадратов отклонений средних групповых сначений от общего среднего 
S_f = np.sum((x_mean_1 - x_mean_total)**2)*8+np.sum((x_mean_2 - x_mean_total)**2)*9+np.sum((x_mean_3 - x_mean_total)**2)*11
S_f

253.9074675324678

In [36]:
# Остаточная сумма квадратов отклонений
S_ost = np.sum((x1 - x_mean_1)**2)+np.sum((x2 - x_mean_2)**2)+np.sum((x3 - x_mean_3)**2)
S_ost

577.0568181818182

In [38]:
S_ = S_f + S_ost
S_ #значение получилось такое же что и s = np.sum((total - x_mean_total)**2)

830.964285714286

Оценки дисперсии

In [39]:
D_f = S_f/(k-1)
D_f

126.9537337662339

In [40]:
D_ost = S_ost/(n-k)
D_ost

23.08227272727273

Значение статистики

In [42]:
Fn = D_f/D_ost
Fn

5.500053450812598

In [43]:
f = stats.f_oneway(x1,x2,x3)
f

F_onewayResult(statistic=5.500053450812596, pvalue=0.010482206918698693)

Далее чтобы сделать вывод есть ли различия в росте среди футболистов, хоккеистов и штангистов, найдем по таблице Фишера табличное значение критерия F
Расчитаем степени свободы - (k-1) = 3-1 = 2    (n-k) = 28-3 = 25     
2  и  25
табличное значение F_crit  = 3,38  

In [44]:
# проверяем через функцию  
alpha = 0.05
k1 = 2
k2 = 25
F_crit = stats.f.ppf(1 - alpha, k1, k2)
F_crit

3.3851899614491683

Видим что f > F_crit (5.500053450812596 > 3.3851899614491683) значит делаем вывод что есть статистически значимые различия среди футболистов, хоккеистов и штангистов

Далее попробуем  воспользоваться post hoc test - ом  Тьюки

In [5]:
import pandas as pd
import numpy as np
from scipy. stats import f_oneway
from statsmodels.stats.multicomp import pairwise_tukeyhsd

df = pd.DataFrame({'score': [173, 175, 180, 178, 177, 185, 183, 182,
                            177, 179, 180, 188, 177, 172, 171, 184, 180,
                            172, 173, 169, 177, 166, 180, 178, 177, 172, 166, 170],
                            'group': np.repeat(['Футболисты', 'Хоккеисты', 'Штангисты'], repeats =(8, 9, 11))})
tukey = pairwise_tukeyhsd(endog=df['score'], groups=df['group'], alpha=0.05)
print(tukey)



    Multiple Comparison of Means - Tukey HSD, FWER=0.05     
  group1     group2  meandiff p-adj   lower    upper  reject
------------------------------------------------------------
Футболисты Хоккеисты  -0.4583  0.979  -6.2732  5.3566  False
Футболисты Штангисты  -6.3977 0.0219 -11.9583 -0.8372   True
 Хоккеисты Штангисты  -5.9394 0.0284 -11.3181 -0.5607   True
------------------------------------------------------------


Из расчетов видно что наиболее ближе значения лежат между футболистами и хоккеистами 
из колонки p-adj  видно  статистически значимые различия 
значения которые меньша 0,05 говорят о наличии статистически значимых различий.